### PR TITLE
langgraph: expand handle_tool_errors in ToolNode

### PR DESCRIPTION
**Pull request by isahers1**
This change expands error-handling functionality of the `ToolNode` by introducing more options for `handle_tool_errors`. Default behavior of the `ToolNode` is unchanged -- all errors are handled and wrapped in a `ToolMessage` to be sent back to LLM.

With this change, users have flexibility to only handle the exceptions that they need to pass back to the LLM:

* they can specify exceptions to handle by passing a tuple of exceptions in `handle_tool_errors`
* specify `handle_tool_errors=True/str/callable`
  * when `handle_tool_errors` is a callable, the signature will be inspected and exceptions from the signature will be handled

### CODE REVIEW

File: libs/langgraph/langgraph/prebuilt/tool_node.py [hunk 1]
```diff
@@ -67,13 +71,83 @@ def msg_content_output(output: Any) -> str | List[dict]:
...
     a list of ToolMessages, one for each tool call.
 
+    Args:
+        tools: A sequence of tools that can be invoked by the ToolNode.
+        name: The name of the ToolNode in the graph. Defaults to "tools".
+        tags: Optional tags to associate with the node. Defaults to None.
+        handle_tool_errors: How to handle tool errors raised by tools inside the node. Defaults to "all".
+            Must be one of the following:
+
+            - "all": all non-validation errors will be caught and
```
Comment (baskaryan) [inline libs/langgraph/langgraph/prebuilt/tool_node.py:130]: why have "all" and True?

File: libs/langgraph/langgraph/prebuilt/tool_node.py [hunk 2]
```diff
@@ -67,13 +71,109 @@ def msg_content_output(output: Any) -> str | List[dict]:
...
 
     It can be used either in StateGraph with a "messages" key (or a custom key passed via ToolNode's 'messages_key').
     If multiple tool calls are requested, they will be run in parallel. The output will be
     a list of ToolMessages, one for each tool call.
 
+    Args:
+        tools: A sequence of tools that can be invoked by the ToolNode.
+        name: The name of the ToolNode in the graph. Defaults to "tools".
+        tags: Optional tags to associate with the node. Defaults to None.
+        handle_tool_errors: How to handle tool errors raised by tools inside the node. Defaults to "all".
```
Comment (eyurtsev) [inline libs/langgraph/langgraph/prebuilt/tool_node.py:153]: defaults to "all" but all is not documented?

File: libs/langgraph/langgraph/prebuilt/tool_node.py [hunk 3]
```diff
@@ -67,13 +71,109 @@ def msg_content_output(output: Any) -> str | List[dict]:
...
+                a ToolMessage with a default error message (TOOL_CALL_ERROR_TEMPLATE) will be returned.
+            - str: all non-validation errors will be caught and
+                a ToolMessage with the string value of 'handle_tool_errors' will be returned.
+            - tuple[type[Exception], ...]: exceptions in the tuple will be caught and
+                a ToolMessage with a default error message (TOOL_CALL_ERROR_TEMPLATE) will be returned.
+            - Callable[..., str]: exceptions from the signature of the callable will be caught and
+                a ToolMessage with the string value of the result of the 'handle_tool_errors' callable will be returned.
+            - False: none of the errors raised by the tools will be caught
+
+        handle_validation_errors: How to handle validation errors raised by tools inside the node. Defaults to True.
```
Comment (eyurtsev) [inline libs/langgraph/langgraph/prebuilt/tool_node.py:166]: This logic assumes that a validation error b/c the tool call arguments do not match the function signature of the tool, but a validation error can also be raised from inside the tool (e.g., a tool that uses pydantic models inside).

Is there a way to consolidate into a parameter maybe that's more generic?

File: libs/langgraph/langgraph/prebuilt/tool_node.py (hunk 2 above)
Comment (vbarda) [reply libs/langgraph/langgraph/prebuilt/tool_node.py:153]: typo, thanks for catching (leftover from previous version)

File: libs/langgraph/langgraph/prebuilt/tool_node.py (hunk 3 above)
Comment (vbarda) [reply libs/langgraph/langgraph/prebuilt/tool_node.py:166]: yea, let's consolidate into `handle_tool_errors`

File: libs/langgraph/tests/test_prebuilt.py [hunk 4]
```diff
@@ -368,6 +374,39 @@ def tool2(some_val: int) -> str:
...
+    expected = (ValueError, ToolException)
+    actual = _infer_handled_types(handle3)
+    assert expected == actual
+
+    expected = (ValueError,)
+    actual = _infer_handled_types(handle4)
+    assert expected == actual
+
+
 async def test_tool_node():
```
Comment (eyurtsev) [inline libs/langgraph/tests/test_prebuilt.py:410]: This is a unit test with 300 lines of stuff in it? Any way to break apart? It's hard to follow right now to understand if the API works as expected

File: libs/langgraph/tests/test_prebuilt.py [hunk 5]
```diff
@@ -394,6 +433,20 @@ async def tool4(some_val: int, some_other_val: str) -> str:
             {"type": "image_url", "image_url": {"url": "abdc"}},
         ]
 
+    @dec_tool
```
Comment (eyurtsev) [inline libs/langgraph/tests/test_prebuilt.py:436]: Does the tool decorator not support `handle_tool_error` as a parameter ?

File: libs/langgraph/tests/test_prebuilt.py [hunk 6]
```diff
@@ -368,6 +374,39 @@ def tool2(some_val: int) -> str:
...
 
 
+def test__infer_handled_types() -> None:
+    def handle(e):  # type: ignore
+        return ""
+
+    def handle2(e: Exception) -> str:
+        return ""
+
+    def handle3(e: Union[ValueError, ToolException]) -> str:
```
Comment (eyurtsev) [inline libs/langgraph/tests/test_prebuilt.py:384]: What happens if the function is misstyped?

```Union[Union[ValueError, TypeError], ToolException]]```

or
```Tuple[ValueError, TypeError]```

**Review by eyurtsev — APPROVED** [review]

**Review by eyurtsev — CHANGES_REQUESTED** [review]
one sec

File: libs/langgraph/langgraph/prebuilt/tool_node.py [hunk 7]
```diff
@@ -67,13 +69,75 @@ def msg_content_output(output: Any) -> str | List[dict]:
...
 
     It can be used either in StateGraph with a "messages" key (or a custom key passed via ToolNode's 'messages_key').
     If multiple tool calls are requested, they will be run in parallel. The output will be
     a list of ToolMessages, one for each tool call.
 
+    Args:
+        tools: A sequence of tools that can be invoked by the ToolNode.
+        name: The name of the ToolNode in the graph. Defaults to "tools".
+        tags: Optional tags to associate with the node. Defaults to None.
+        handle_tool_errors: How to handle tool errors raised by tools inside the node. Defaults to True.
```
Comment (eyurtsev) [inline libs/langgraph/langgraph/prebuilt/tool_node.py:128]: Is `True` the safe default? I know it's convenient, but what's the contents of `error` in the error message?

Is it the full stack trace? Just the error message? Could something sensitive be leaked by mistake as part of the error message? (e.g., API Key)

File: libs/langgraph/langgraph/prebuilt/tool_node.py [hunk 8]
```diff
@@ -67,13 +69,75 @@ def msg_content_output(output: Any) -> str | List[dict]:
...
+            Must be one of the following:
+
+            - True: all errors will be caught and
+                a ToolMessage with a default error message (TOOL_CALL_ERROR_TEMPLATE) will be returned.
+            - str: all errors will be caught and
```
Comment (eyurtsev) [inline libs/langgraph/langgraph/prebuilt/tool_node.py:133]: Does string support a template variable? {error} ?

Comment (vbarda) [reply libs/langgraph/langgraph/prebuilt/tool_node.py:133]: Nope, i think at that point users should just use callable. do you think we should remove?

File: libs/langgraph/langgraph/prebuilt/tool_node.py (hunk 7 above)
Comment (vbarda) [reply libs/langgraph/langgraph/prebuilt/tool_node.py:128]: it's the current default, so i think it would be problematic to change it as it would be breaking. the content is just the error message, no stack trace

hm, that's an interesting point about a sensitive info in error message -- possible? but IMO this is the same class of issue as "what if someone invokes a tool/graph with their API key as input"

File: libs/langgraph/tests/test_prebuilt.py (hunk 6 above)
Comment (vbarda) [reply libs/langgraph/tests/test_prebuilt.py:384]: will break -- good catch

Comment (vbarda) [reply libs/langgraph/tests/test_prebuilt.py:384]: added a better error message for those situations

File: libs/langgraph/langgraph/prebuilt/tool_node.py (hunk 7 above)
Comment (eyurtsev) [reply libs/langgraph/langgraph/prebuilt/tool_node.py:134]: if it's default can leave as is

File: libs/langgraph/langgraph/prebuilt/tool_node.py (hunk 8 above)
Comment (eyurtsev) [reply libs/langgraph/langgraph/prebuilt/tool_node.py:139]: I think it's fine either way

File: libs/langgraph/langgraph/prebuilt/tool_node.py [hunk 9]
```diff
@@ -67,13 +69,81 @@ def msg_content_output(output: Any) -> str | List[dict]:
...
+        if params[0].name in ["self", "cls"] and len(params) == 2:
+            first_param = params[1]
+        else:
+            first_param = params[0]
+
+        type_hints = get_type_hints(handler)
+        if first_param.name in type_hints:
+            origin = get_origin(first_param.annotation)
+            if origin is Union:
+                return tuple(get_args(first_param.annotation))
```
Comment (eyurtsev) [inline libs/langgraph/langgraph/prebuilt/tool_node.py:110]: What if it's 

`Union[Tuple[str, str]]` ?

File: libs/langgraph/langgraph/prebuilt/tool_node.py [hunk 10]
```diff
@@ -67,13 +69,81 @@ def msg_content_output(output: Any) -> str | List[dict]:
...
+            elif origin is not None:
+                raise ValueError(
+                    "Generic types are not supported in the error handler signature. "
```
Comment (eyurtsev) [inline libs/langgraph/langgraph/prebuilt/tool_node.py:113]: Generic is a thing already

```suggestion
                    "Arbitrary types are not supported in the error handler signature."
```

File: libs/langgraph/langgraph/prebuilt/tool_node.py [hunk 11]
```diff
@@ -67,13 +69,81 @@ def msg_content_output(output: Any) -> str | List[dict]:
...
+                    "Please use only Exception types or a union of Exception types."
```
Comment (eyurtsev) [inline libs/langgraph/langgraph/prebuilt/tool_node.py:114]:
```suggestion
                    "Please annotate the error with either a specific Exception type or a union of Exception types. "
```

File: libs/langgraph/langgraph/prebuilt/tool_node.py [hunk 12]
```diff
@@ -67,13 +69,81 @@ def msg_content_output(output: Any) -> str | List[dict]:
...
+                )
```
Comment (eyurtsev) [inline libs/langgraph/langgraph/prebuilt/tool_node.py:115]: If you like could do something like this to be even more helpful:

```suggestion
   "For example, `def custom_handler(e: ValueError)` or `def custom-handler(e: Union[ValueError, TypeError])`
                )
```

and could also try to output the parameter signature

**Review by eyurtsev — APPROVED** [review]

File: libs/langgraph/langgraph/prebuilt/tool_node.py (hunk 9 above)
Comment (vbarda) [reply libs/langgraph/langgraph/prebuilt/tool_node.py]: added more restrictive checks & better error messages
